booking/src/main/java/org/dcsa/conformance/standards/booking/party/BookingShipper.java

```diff
@@ -109,35 +109,32 @@ private void sendBookingRequest(JsonNode actionPrompt) {
   }
 
   private JsonNode replaceBookingPlaceHolders(JsonNode actionPrompt) {
-
     CarrierScenarioParameters carrierScenarioParameters =
       CarrierScenarioParameters.fromJson(actionPrompt.get("csp"));
     var scenarioType = ScenarioType.valueOf(actionPrompt.required("scenarioType").asText());
+    var replacements = new HashMap<String, String>();
+    replacements.put(
+        "SERVICE_CONTRACT_REFERENCE_PLACEHOLDER",
+        carrierScenarioParameters.serviceContractReference());
+    replacements.put(
+        "CONTRACT_QUOTATION_REFERENCE_PLACEHOLDER",
+        carrierScenarioParameters.contractQuotationReference());
+    replacements.put(
+        "CARRIER_EXPORT_VOYAGE_NUMBER_PLACEHOLDER",
+        carrierScenarioParameters.carrierExportVoyageNumber());
+    replacements.put(
+        "CARRIER_SERVICE_NAME_PLACEHOLDER", carrierScenarioParameters.carrierServiceName());
+    replacements.put("COMMODITY_HS_CODE_1", carrierScenarioParameters.hsCodes1());
+    replacements.put("COMMODITY_HS_CODE_2", carrierScenarioParameters.hsCodes1());
+    replacements.put("COMMODITY_TYPE_1_PLACEHOLDER", carrierScenarioParameters.commodityType1());
+    replacements.put("COMMODITY_TYPE_2_PLACEHOLDER", carrierScenarioParameters.commodityType2());
+    replacements.put(
+        "POL_UNLOCATION_CODE_PLACEHOLDER", carrierScenarioParameters.polUNLocationCode());
+    replacements.put(
+        "POD_UNLOCATION_CODE_PLACEHOLDER", carrierScenarioParameters.podUNLocationCode());
+
     return JsonToolkit.templateFileToJsonNode(
-        "/standards/booking/messages/"+ scenarioType.bookingTemplate(apiVersion),
-        Map.ofEntries(
```

Comment:
Actually, nothing is wrong with the existing way of writing that. But since I changed the mapper, some values could be `null`, which is not allowed if you use `Map.entry()`. So, I favored to use the generic way of mapping objects and I needed to change this code here, to prevent failures. 

Next to that, I changed `JsonToolkit.templateFileToJsonNode` method, to get the same replacing behavior as before, because developers used `cspNode.required("commodityType1").asText(),` way to create the Java object, which causes the String objects to get filled with "null" text.  
Actually, I am now in doubt if I really need both changes. Will check if the null-check is really needed later (before merging).

Reply:
I refactored it a bit, so now it uses the `Objects.requireNonNullElse()` method, so we can keep the fluent API and be null safe. Also reverted  the change in `JsonToolkit.templateFileToJsonNode` method.